### PR TITLE
resholve: mark it as knownVulnerabilities, allow resholve-utils usage

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Description of changes

We are marking `resholve` itself with `meta.knownVulnerabilities`, and overriding `resholve-utils` functions's `resholve` with `meta.knownVulnerabilities = [ ]`.

This way, we can still use `resholve` at build-time without triggering security warnings, however we can't instantiate `resholve` itself. See:

```
$ nix-build -A resholve
error: Package ‘resholve-0.8.4’ in /.../nixpkgs/pkgs/development/misc/resholve/resholve.nix:48 is marked as insecure, refusing to evaluate.

$ nix-build -A ix
/nix/store/k8cvj1bfxkjj8zdg6kgm7r8942bbj7w7-ix-20190815
```

For debugging purposes, you can still bypass the security checks and instantiate `resholve` by:

```
$ NIXPKGS_ALLOW_INSECURE=1 nix-build -A resholve
/nix/store/77s87hhqymc6x9wpclb04zg5jwm6fsij-resholve-0.8.4
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 210078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>arch-install-scripts</li>
    <li>bash-preexec</li>
    <li>bashup-events32</li>
    <li>bashup-events44</li>
    <li>bats</li>
    <li>dgoss</li>
    <li>git-ftp</li>
    <li>ix</li>
    <li>mons</li>
    <li>msmtp</li>
    <li>packcc</li>
    <li>pdf2odt</li>
    <li>shunit2</li>
    <li>yadm</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 210078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>arch-install-scripts</li>
    <li>bash-preexec</li>
    <li>bashup-events32</li>
    <li>bashup-events44</li>
    <li>bats</li>
    <li>dgoss</li>
    <li>git-ftp</li>
    <li>ix</li>
    <li>mons</li>
    <li>msmtp</li>
    <li>packcc</li>
    <li>pdf2odt</li>
    <li>shunit2</li>
    <li>yadm</li>
  </ul>
</details>

File: pkgs/development/misc/resholve/default.nix [hunk 1]
```diff
@@ -98,6 +43,8 @@ rec {
   # funcs to validate and phrase invocations of resholve
   # and use those invocations to build packages
   resholve-utils = callPackage ./resholve-utils.nix {
-    inherit resholve;
+    # we can still use resholve-utils without triggering a security warn
+    # this is safe since we will only use `resholve` at build time
+    resholve = removeKnownVulnerabilities resholve;
```
Comment (dotlambda) [inline pkgs/development/misc/resholve/default.nix:48]: Can't we do something like
```suggestion
    resholve = resholve.override {
       python2 = removeKnownVulnerabilities python2;
     };
```

File: pkgs/development/misc/resholve/resholve-utils.nix [hunk 2]
```diff
@@ -181,6 +181,7 @@ rec {
       src = unresholved;
       inherit version pname;
       buildInputs = [ resholve ];
+      disallowedReferences = [ resholve ];
```
Comment (thiagokokada) [inline pkgs/development/misc/resholve/resholve-utils.nix:184]: Setting this here to make sure that we have no runtime references to `resholve` itself.

**Comment by thiagokokada** [comment]
@ofborg build arch-install-scripts ix mons bats shunit2

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 210078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>arch-install-scripts</li>
    <li>bash-preexec</li>
    <li>bashup-events32</li>
    <li>bashup-events44</li>
    <li>bats</li>
    <li>dgoss</li>
    <li>git-ftp</li>
    <li>ix</li>
    <li>mons</li>
    <li>msmtp</li>
    <li>packcc</li>
    <li>pdf2odt</li>
    <li>shunit2</li>
    <li>yadm</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
@ofborg build arch-install-scripts ix mons bats shunit2

File: pkgs/development/misc/resholve/default.nix [hunk 3]
```diff
@@ -94,10 +39,16 @@ rec {
     inherit (deps.oil) oildev;
     inherit (deps) configargparse;
     inherit resholve-utils;
+    # used only in tests
+    resholveForTests = resholveForUtils;
```
Comment (thiagokokada) [inline pkgs/development/misc/resholve/default.nix:43]: Now that I am thinking about, probably just using `resholve = resholveForUtils` should work.

I will fix this later.

Comment (abathur) [reply pkgs/development/misc/resholve/default.nix:43]: I like the fix for the CLI tests broadly (and I agree that `resholve = resholveForUtils` sounds like an improvement).

File: pkgs/development/misc/resholve/resholve.nix [hunk 4]
```diff
@@ -50,5 +51,6 @@ python27.pkgs.buildPythonApplication {
     license = with licenses; [ mit ];
     maintainers = with maintainers; [ abathur ];
     platforms = platforms.all;
+    knownVulnerabilities = [ "resholve depends on python27 that is EOL" ];
```
Comment (abathur) [inline pkgs/development/misc/resholve/resholve.nix:54]:
```suggestion
    knownVulnerabilities =  [ ''
      resholve depends on python27 (EOL). While it's safe to
         run on trusted input in the build sandbox, you should
         avoid running it on untrusted input.
    '' ];
```

Would like this to be specific enough for people who aren't up on the python2 situation evaluate the risk well enough to decide whether or not to permit it. WDYT?

**Review by abathur — COMMENTED** [review]
Generally on board. Thanks for getting the CLI tests working :)

File: pkgs/development/misc/resholve/default.nix (hunk 3 above)
Comment (thiagokokada) [reply pkgs/development/misc/resholve/default.nix]: Needed to pass `resholve` as argument for `resholve.nix`, but it is working fine with the change.

**Comment by thiagokokada** [comment]
@ofborg build arch-install-scripts ix mons bats shunit2

**Comment by thiagokokada** [comment]
Can you re-review @abathur ?

**Review by abathur — APPROVED** [review]
Looks good. I like both attr renames in the latest change (and that doing so reduced the overall diff by a file :)

Thank you.

Result of `nixpkgs-review pr 210078` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>resholve</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>bash-preexec</li>
    <li>bashup-events32</li>
    <li>bashup-events44</li>
    <li>bats</li>
    <li>git-ftp</li>
    <li>ix</li>
    <li>mons</li>
    <li>msmtp</li>
    <li>packcc</li>
    <li>pdf2odt</li>
    <li>shunit2</li>
    <li>yadm</li>
  </ul>
</details>

**Pull request merged**